### PR TITLE
InputElement class

### LINKED ISSUE #248: `NumericInput` inherits from `TextInput`
`NumericInput` inherits from `TextInput`. This is awkward from a types perspective because a `TextInput` has a value of type `string` but a `NumericInput` has a value of type `number`. 

A better inheritance hierarchy would be:

```
Element
  |--InputElement (abstract base class that creates + manages the input DOM element)
       |--TextInput
       |--NumericInput
```

...where each `TextInput` and `NumericInput` reads/validates/types the actual textual entry of the input element.

### PR DESCRIPTION
**Pull request by ellthompson**
Creates the `InputElement` class which both `TextInput` and `NumericInput` can extend. This allows us to tighten up the value property type of each class (string to `TextInput` and number for `NumericInput`). It also removes all text validation related functionality from the `NumericInput`, which was only ever necessary for the `TextInput`.

Fixes #248

### CODE REVIEW

**Comment by willeastcott** [comment]
Also, just throwing ideas out here, but would `InputElement` be a better name than `InputField` (since it's a specialization of `Element` and we don't really use the 'field` terminology except for in label group).

File: src/components/InputElement/index.ts [hunk 1]
```diff
@@ -0,0 +1,267 @@
+import Element, { ElementArgs, IBindable, IBindableArgs, IFocusable, IPlaceholder, IPlaceholderArgs } from '../Element';
+import * as pcuiClass from '../../class';
+
+const CLASS_INPUT_FIELD = 'pcui-input-element';
```
Comment (willeastcott) [inline src/components/InputElement/index.ts:4]:
```suggestion
const CLASS_INPUT_ELEMENT = 'pcui-input-element';
```

File: src/components/InputElement/index.ts [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+     */
+    keyChange?: boolean,
+    /**
+     * The input element to associate this {@link InputElement} with. If not supplied one will be created instead.
+     */
+    input?: HTMLInputElement
+}
+
+/**
+ * The InputElement is an input element with a text field.
```
Comment (willeastcott) [inline src/components/InputElement/index.ts:29]:
```suggestion
* The InputElement is an input element with a text field. It is an abstract class containing the
* functionality required by all elements that use an input DOM element such as {@link TextInput}
* and {@link NumericInput}. It is not intended to be used directly.
```

File: src/components/InputElement/index.ts [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    protected _blurOnEscape: boolean;
+
+    protected _onInputKeyDownEvt: (evt: KeyboardEvent) => void;
+
+    protected _onInputChangeEvt: (evt: Event) => void;
+
+    constructor(args: InputElementArgs = {}) {
+        super(args);
+
+        this.class.add(CLASS_INPUT_FIELD);
```
Comment (willeastcott) [inline src/components/InputElement/index.ts:53]:
```suggestion
        this.class.add(CLASS_INPUT_FIELD);
```

File: src/components/NumericInput/index.ts [hunk 4]
```diff
@@ -302,8 +299,8 @@ class NumericInput extends TextInput {
         }
     }
 
-    get value() {
-        const val = super.value;
+    get value() : number {
+        const val = this._domInput.value;
         // @ts-ignore
```
Comment (willeastcott) [inline src/components/NumericInput/index.ts:304]: Is this @ts-ignore needed any more?

**Comment by willeastcott** [comment]
Worth marking this PR as `[BREAKING]`? And listing what changes a user might have to make to their app code? I'm thinking of any changes to CSS class names but maybe there's other things to list too? Not sure...

File: src/components/NumericInput/index.ts [hunk 5]
```diff
@@ -86,7 +86,7 @@ class NumericInput extends TextInput {
         this._min = args.min ?? null;
         this._max = args.max ?? null;
         this._allowNull = args.allowNull ?? false;
-        this._precision = args.precision ?? 7;
+        this._precision = args.precision ?? 2;
```
Comment (willeastcott) [inline src/components/NumericInput/index.ts:89]: This looks like it fixes something else. I'm wondering if this should be a separate PR?

File: src/components/NumericInput/index.ts (hunk 4 above)
Comment (ellthompson) [reply src/components/NumericInput/index.ts:304]: I'll see if I can remove this one.

**Comment by ellthompson** [comment]
> Worth marking this PR as `[BREAKING]`? And listing what changes a user might have to make to their app code? I'm thinking of any changes to CSS class names but maybe there's other things to list too? Not sure...

The functionality of NumericInput and TextInput shouldn't change. They'll both still have the same class names attached to them so it shouldn't break end users css, it'll now also include an additional `pcui-input-element` class. The only thing I can think of which could break is if users are building their own version of the styling using the distributed scss files, as some scss rules would have their names changed. That process has never been documented publicly though.
